File: src/RepoLite/RepoLite.GeneratorEngine/Templates/CSharp/BaseModel.cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace MODELNAMESPACE.Base
{
    public class ValidationError
    {
        public string PropertyName { get; set; }
        public string Error { get; set; }

        public ValidationError(string property, string error)
        {
            PropertyName = property;
            Error = error;
        }
    }

    public abstract class BaseModel
    {
        public abstract List<ValidationError> Validate();
        public readonly List<string> DirtyColumns = new List<string>();

        public void ResetDirty()
        {
            DirtyColumns.Clear();
        }

        protected void SetValue<T>(ref T prop, T value, [CallerMemberName] string propName = "")
        {
            if (!DirtyColumns.Contains(propName))
            {
                DirtyColumns.Add(propName);
            }
            prop = value;
        }

        public static int GetDecimalPlaces(decimal n)
        {
            n = Math.Abs(n); //make sure it is positive.
            n -= (int)n;     //remove the integer part of the number.
            var decimalPlaces = 0;
            while (n > 0)
            {
                decimalPlaces++;
                n *= 10;
                n -= (int)n;
            }
            return decimalPlaces;
        }
    }
}

File: src/RepoLite.VB.Tests/ActualGeneratedFileTests/GeneratedFiles/Base/BaseModel.vb
Imports System.Runtime.CompilerServices

Namespace MODELNAMESPACE.Base
    Public Class ValidationError
        Public Property PropertyName As String
        Public Property [Error] As String

        Public Sub New([property] As String, [error] As String)
            PropertyName = [property]
            Me.[error] = [error]
        End Sub
    End Class

    Public MustInherit Class BaseModel
        Public MustOverride Function Validate() As List(Of ValidationError)
        Public ReadOnly DirtyColumns As List(Of String) = New List(Of String)()

        Public Sub ResetDirty()
            DirtyColumns.Clear()
        End Sub

        Protected Sub SetValue(Of T)(ByRef prop As T, value As T,
                                      <CallerMemberName> Optional propName As String = "")
            If Not DirtyColumns.Contains(propName) Then
                DirtyColumns.Add(propName)
            End If

            prop = value
        End Sub

        Public Shared Function GetDecimalPlaces(n As Decimal) As Integer
            n = Math.Abs(n)
            n -= CInt(n)
            Dim decimalPlaces = 0

            While n > 0
                decimalPlaces += 1
                n *= 10
                n -= CInt(n)
            End While

            Return decimalPlaces
        End Function
    End Class
End Namespace
